packages/core/src/modules/credentials/formats/indy/IndyCredentialFormatService.ts

```diff
@@ -0,0 +1,750 @@
+          offer.indy.payload.credentialPayload?.cred_def_id
+        )
+        return {
+          indy: {
+            credDef: indyCredDef,
+          },
+        }
+      }
+    }
+  }
+
+  /**
+   * Get linked attachments for indy format from a proposal message. This allows attachments
+   * to be copied across to old style credential records
+   *
+   * @param proposal ProposeCredentialOptions object containing (optionally) the linked attachments
+   * @return array of linked attachments or undefined if none present
+   */
+  private getCredentialLinkedAttachments(proposal: ProposeCredentialOptions): {
+    attachments: Attachment[] | undefined
+    previewWithAttachments: V2CredentialPreview
+  } {
+    // Add the linked attachments to the credentialProposal
+    const credPropose: CredPropose = proposal.credentialFormats.indy?.payload.credentialPayload as CredPropose
+
+    let attachments: Attachment[] | undefined
+    let previewWithAttachments: V2CredentialPreview = new V2CredentialPreview({
+      attributes: credPropose.attributes ? credPropose.attributes : [],
+    })
+    if (proposal.credentialFormats.indy && credPropose.linkedAttachments) {
+      // there are linked attachments so transform into the attribute field of the CredentialPreview object for
+      // this proposal
+      if (credPropose.attributes && credPropose.credentialDefinitionId) {
+        previewWithAttachments = CredentialUtils.createAndLinkAttachmentsToPreview(
+          credPropose.linkedAttachments,
+          new V2CredentialPreview({
+            attributes: credPropose.attributes,
+          })
+        )
+      }
+      attachments = credPropose.linkedAttachments.map((linkedAttachment) => linkedAttachment.attachment)
+
+      credPropose.credentialDefinitionId = this.getCredentialDefinitionId(proposal)
+
+      proposal.credentialFormats.indy.payload.credentialPayload = credPropose
+    }
+    return { attachments, previewWithAttachments }
+  }
+
+  /**
+   *
+   * @param options Gets the credential definition id if present for an indy credential
+   * @returns the credential definition id for this credential
+   */
+  private getCredentialDefinitionId(options: ProposeCredentialOptions): string | undefined {
+    const credPropose: CredPropose = options.credentialFormats.indy?.payload.credentialPayload as CredPropose
+    return credPropose.credentialDefinitionId
+  }
+  /**
+   * Get attributes for indy format from a proposal message. This allows attributes
+   * to be copied across to old style credential records
+   *
+   * @param proposal ProposeCredentialOptions object containing (optionally) the attributes
+   * @return array of attributes or undefined if none present
+   */
+  public getCredentialAttributes(proposal: ProposeCredentialOptions): CredentialPreviewAttribute[] | undefined {
+    const credPropose: CredPropose = proposal.credentialFormats.indy?.payload as CredPropose
```

Comment:
With the more strict types we wouldn't have to cast everytime. Looser typing can easily lead to bugs! (Typescript won't complain when the type changes)